### DISCUSSION TITLE
Importing external skeleton data or centerline graph

### DISCUSSION THREAD

**Discussion by lukasglandorf** [JacobBumgarner/VesselVio#63] [Q&A]
Hi, great work on the package!
I would like to use VesselVio for 3D visualization and need to import my data for this. Specifically, I'm interested in supplying my own skeleton/centerlines and visualizing them as tubes + some additional vessel properties that are not included in VesselVio. My external analysis pipeline is also graph-based so I can also import a graph.

My question is if there's already some functionality that enables this or if I have to write my own import functions. I found references to a "centerline graph" in the code. Do you have an example for the dataformat? If it's just vertices,edges and some additional attribute that specifies the centerline voxels, that would be exactly what I'm looking for.

(Adding some code to read additional fields for my own vessel properties will not be a problem. If you're interested I might consider to make a pull request to allow reading arbitrary additional attributes as input that can be visualized by the colormap).

**Comment by JacobBumgarner** [comment]
Hey there @lukasglandorf. Thanks!

You should be able to supply your own graph data to the app and visualize them. What is the format of your graph? You can either supply CSV files or a variety of different graph files. An example CSV file input can be found here:
https://github.com/JacobBumgarner/VesselVio/tree/main/sample_data/demo_csv

**Reply by lukasglandorf** [reply]
Thanks for the quick reply!

I'm importing my graph from Voreen (https://www.uni-muenster.de/Voreen/) because I prefer their skeletonization algorithm. The csv files for edges and vertices are identical to yours, so edge and vertex attributes are not a problem. I'm then doing some further processing in Matlab. So I plan to export from Matlab to VesselVio. I think the only thing I'm looking for is how to import explicit skeleton voxels for each edge. So basically a list of 3D coordinates for each edge. I'm not sure the csv approach supports that out of the box?
However, I've just found what looks like you provide each edge with a `coords_list` attribute (refer line 224 in `volume_visualization.py`). That sounds like it's exactly what I'm looking for?

**Reply by JacobBumgarner** [reply]
Yes - once the graph has been 'reduced' at the end of the analysis process, all of the segment vertices are condensed into a single edge, and the coordinate information from the segment vertices is stored in that `coords_list` attribute. 

Unfortunately, at least when I was experimenting with graph file exports when I originally made the app, one could not export lists/arrays as attributes of graph edges, so I didn't have that as an importable feature. Has this changed recently?
